### ISSUE TITLE
GET requests to :4444/graphql return Forbidden

### ISSUE THREAD

**Issue by bogdanbiv** [codefoundries/isomorphic-material-relay-starter-kit#190]
1) Unauthenticated GET http://localhost:4444/graphql results in 403 (Forbidden)
eyJ1c2VyX2lkIjoiYmVlMmZkNDMtYjYxZS00YWRkLTg0MmMtODdjZWEyMDlkMDcwIn0.8_7uV3G1W7sEfkFoQgzuhWt

From the server's output in console:
warn: Checking credentials failed: Authentication token expected: 802MVIJ9tymWeaNHUqLRfuGsteTOI3vhdhk31s902BT8J0Lta9rNHT9MNVaBfigd9s2974h4RfuGVtbh32o902BT8Jiew29uT913sHU, provided: **undefined**
No cookie is set in the browser at this point

2) Also, I believe this is the server log for the GET request http://localhost:4444/graphql  authenticated with a fresh user:
warn: Checking credentials failed: Authentication token expected: 0.eu519m21qol84cxr, provided: **undefined**
In this case a *auth_token* cookie is set as a jwt which decodes to { "typ": "JWT", "alg": "HS256" }
{ "user_id": "bee2fd43-b61e-4add-842c-87cea209d070" }

$ cat .env
PORT=4444
HOST=127.0.0.1
JWT_SECRET=QVN...too_long....
ANONYMOUS_USER_AUTH_TOKEN=802MVIJ9tymWeaNHUqLRfuGsteTOI3vhdhk31s902BT8J0Lta9rNHT9MNVaBfigd9s2974h4RfuGVtbh32o902BT8Jiew29uT913sHU
OBJECT_PERSISTENCE=memory
CASSANDRA_CONNECTION_POINTS=localhost
CASSANDRA_KEYSPACE=imrsk

$ npm -v   -- 3.7.3
$ node -v  -- v5.8.0
$ cassandra -v -- 2.0.11
$ cqlsh 4.1.1

$ javac -version --  javac 1.8.0_77
$ java -version
java version "1.8.0_77"
Java(TM) SE Runtime Environment (build 1.8.0_77-b03)
Java HotSpot(TM) 64-Bit Server VM (build 25.77-b03, mixed mode)

Ubuntu 15.10

**Comment by thelordoftheboards** [comment]
There are two tokens at play - one is the JWT token in HTTP only cookie protecting the app against XSS and a token passed from  the user record through GraphQL to protect against the CRSF. It seems that the latter might be a problem. I will add something to print it for debugging purposes in a few minutes.

**Comment by thelordoftheboards** [comment]
I added this on the bottom of the home screen:

![image](https://cloud.githubusercontent.com/assets/7056535/14071830/376f7ba2-f46e-11e5-91e4-be55d0db2b7d.png)

What does it display for you? If you use Cassandra persistence, it should be easy to verify it matches  the record there:

![image](https://cloud.githubusercontent.com/assets/7056535/14071851/71c7249e-f46e-11e5-9c04-9afeafce38c8.png)

**Comment by bogdanbiv** [comment]
Could you please clarify where should I put the debugging instructions which print CSRF / User Auth tokens?

I can confirm cassandra backend is working properly:
```
 id                                   | User_AccountName          | User_AccountPassword                                         | User_AuthToken     | User_DisplayName | User_Email | User_Locale | User_ProfilePhoto
--------------------------------------+---------------------------+--------------------------------------------------------------+--------------------+------------------+------------+-------------+-------------------
 5d92ac17-5d91-4088-8592-0b47d110e5a5 | b*********@*****.com | $2a$08$vTYy8uOU2wALN09VRln85uR6.yK0cyCG09W7Xd3oYOhfF5HAuTBt6 | 0.gqvlcwniao979zfr |         New User |          
```

**Comment by thelordoftheboards** [comment]
Sorry, I did not realize I only added it to develop. Check this out:

https://github.com/codefoundries/isomorphic-material-relay-starter-kit/commit/3b15d927c863a93d8aa0051ea5cc2d7cdaa80200

If you add it you should see `0.gqvlcwniao979zfr` at the bottom of the scren.

**Comment by thelordoftheboards** [comment]
I re-read the issue and I now understand I did not get it initially. I was assuming the request was from the app itself rather than the UI. Now I get it.

**Comment by bogdanbiv** [comment]
Checked out 3b15d92, I do see "0.gqvlcw...fr" at the bottom as you indicated.
Just to be clear - my problem so far is that I cannot access the "GraphiQL browser" URL http://localhost:4444/graphql - server responds with 403.

Logs loaded into PasteBin.com - `http://pastebin.com/vgbEmnHj`. 
 * [1] warn: Checking credentials failed: Authentication token expected: 0.gqvlcwniao979zfr, provided:undefined
 * [1] warn: Checking credentials failed: Authentication token expected: 802MVIJ9tymWeaNHUqLRfuGsteTOI3vhdhk31s902BT8J0Lta9rNHT9MNVaBfigd9s2974h4RfuGVtbh32o902BT8Jiew29uT913sHU, provided:undefined

Also:
`[1] info: Starting application npm_package_name=isomorphic-material-relay-starter-kit, npm_package_version=1.0.4, NODE_ENV=development, HOST=127.0.0.1, PORT=4444, PUBLIC_URL=undefined, process_title=/usr/local/bin/node, process_pid=3727, objectPersistence=cassandra, CASSANDRA_KEYSPACE=imrsk, CASSANDRA_CONNECTION_POINTS=localhost          `  
[0] /home/bogdanbiv/Projects/imrsk/node_modules/sockjs/lib/trans-websocket.js:157
[0]       return this.session.close(3000, 'No response from heartbeat');
[0]                          ^
[0]
[0] TypeError: Cannot read property 'close' of null
[0]     at WebSocketReceiver.heartbeat_timeout (/home/bogdanbiv/Projects/imrsk/node_modules/sockjs/lib/trans-websocket.js:157:26)

**Comment by thelordoftheboards** [comment]
I got it now, initially I did not understand that it was direct call, I was assuming it was a call from the app.
Not quite sure what to do with it right now. I tried back in the day to embed the GraphiQL object inside a page in the app, but server rendering would fail. I will have to look at what options for authentication can be passed. Or maybe check again if embedding in a page can be done.
I will also look into how to temporarily disable the check to be able to use GraphiQL
